### PR TITLE
🐛 Source Jira: returns config_error after low-code migration without reset

### PR DESCRIPTION
**Pull request by maxi297**
## What
Addresses https://github.com/airbytehq/airbyte-internal-issues/issues/8257

The goal is to improve our visibility on errors that are actually unexpected and there we need to take action. All the cases we've seen for this error is users not doing the proper migration steps.

## How
Updating the CDK version

## User Impact
This error should now be reported as a config error and therefore improve the accuracy of our error reporting. For the user that is trying to sync the data, it shouldn't not change anything

## Can this PR be safely reverted and rolled back?

- [X] YES 💚
- [ ] NO ❌

### CODE REVIEW

**Review by bazarnov — APPROVED** [review]

File: airbyte-integrations/connectors/source-jira/unit_tests/test_components.py
Comment (maxi297) [inline airbyte-integrations/connectors/source-jira/unit_tests/test_components.py:117]: This test would not pass because `fields_parent_stream_config.parent_key` was a mock and therefore `slices` would be and empty array. I've updated the test with the code on master and ported it here without any change

File: airbyte-integrations/connectors/source-jira/source_jira/streams.py
Comment (darynaishchenko) [inline airbyte-integrations/connectors/source-jira/source_jira/streams.py:352]: @brianjlai Could you please advise me with rfr?

I updated jira to v^3 airbyte-cdk and faced with the issue when python streams stuck because `next_slice` is never None to exit `while` loop [here](https://github.com/airbytehq/airbyte/blob/master/airbyte-cdk/python/airbyte_cdk/sources/streams/core.py#L177). 
`next_slice={}`
When I'm setting `is_resumable = False` stream works as expected.

What am I missing here?

File: airbyte-integrations/connectors/source-jira/source_jira/utils.py
Comment (maxi297) [inline airbyte-integrations/connectors/source-jira/source_jira/utils.py:32]: What is this code doing differently than `entrypoint_wrapper.read`? It feels like it would be less code to maintain and therefore better to use this

File: airbyte-integrations/connectors/source-jira/source_jira/streams.py
Comment (maxi297) [inline airbyte-integrations/connectors/source-jira/source_jira/streams.py:48]: It seems like this would raise an unexpected exception if `response_or_exception` is of type `Exception`. Should we handle this?

File: airbyte-integrations/connectors/source-jira/source_jira/streams.py
Comment (maxi297) [inline airbyte-integrations/connectors/source-jira/source_jira/streams.py:32]: Should we have unit tests for that?

File: airbyte-integrations/connectors/source-jira/source_jira/manifest.yaml
Comment (maxi297) [inline airbyte-integrations/connectors/source-jira/source_jira/manifest.yaml:41]: Why did we need to add this and can't rely on the interpolation of `{{ parameters.extract_field }}` anymore?

**Review by maxi297 — COMMENTED** [review]
I added a couple of questions and I am wondering what was the outcome of [this comment](https://github.com/airbytehq/airbyte/pull/39558#discussion_r1695592075)

File: airbyte-integrations/connectors/source-jira/source_jira/streams.py
Comment (maxi297) [reply airbyte-integrations/connectors/source-jira/source_jira/streams.py]: @brianjlai any insight on this one?

File: airbyte-integrations/connectors/source-jira/metadata.yaml
Comment (maxi297) [inline airbyte-integrations/connectors/source-jira/metadata.yaml:13]: Based on [this discussion](https://airbytehq-team.slack.com/archives/C02U9R3AF37/p1722355757502729), we should decide which version we want to set here

File: airbyte-integrations/connectors/source-jira/source_jira/streams.py
Comment (darynaishchenko) [reply airbyte-integrations/connectors/source-jira/source_jira/streams.py]: Thanks, updated

File: airbyte-integrations/connectors/source-jira/source_jira/streams.py
Comment (darynaishchenko) [reply airbyte-integrations/connectors/source-jira/source_jira/streams.py:32]: added unit test for 400 and 403 errors

File: airbyte-integrations/connectors/source-jira/source_jira/manifest.yaml
Comment (darynaishchenko) [reply airbyte-integrations/connectors/source-jira/source_jira/manifest.yaml:41]: It's due to changes in `JsonDecoder`. Before decoder used to return all json response like: `[{entity 1}, {entity 2}]`. So to extract records path ["*"] were used. But now records are being yielded one by one from response by `JsonDecoder`. So path ["*"] works incorrectly in this case and we just need to keep path empty to return record from JsonDecoder

Comment (maxi297) [reply airbyte-integrations/connectors/source-jira/source_jira/manifest.yaml:41]: I think you are referring to [this PR](https://github.com/airbytehq/airbyte/pull/38829) right? When checking this one, it seems like [the * parameter](https://github.com/airbytehq/airbyte/pull/38829/files#diff-54c832113b872fcf5b9a42fffcf84e45fe2c9e5384cd9af012b84648a23ba747R73-R74) is still supported. From what you mention though, the behavior has changed, is that correct? Can you share a snippet of code or a test that I can perform on my side to understand this change?

File: airbyte-integrations/connectors/source-jira/source_jira/utils.py
Comment (maxi297) [reply airbyte-integrations/connectors/source-jira/source_jira/utils.py:32]: I thought this was for testing purposes. Ignore this one. I think what we want is `Stream.read_only_records`. Would that make more sense?

File: airbyte-integrations/connectors/source-jira/source_jira/manifest.yaml
Comment (darynaishchenko) [reply airbyte-integrations/connectors/source-jira/source_jira/manifest.yaml:41]: I'm referring to [this](https://github.com/airbytehq/airbyte/pull/38829/files#diff-404f304b08b4a327259c7999f75283e1f111edaac2bb09196cb2c536f141f2deR32) change in json decoder class. 
We don't need` "*" ` anymore, as dpath receives record in the loop [here](https://github.com/airbytehq/airbyte/pull/38829/files#diff-54c832113b872fcf5b9a42fffcf84e45fe2c9e5384cd9af012b84648a23ba747R68). 
You can reproduce it by adding` "*" `in base selector and related streams will fail.

Comment (maxi297) [reply airbyte-integrations/connectors/source-jira/source_jira/manifest.yaml:41]: Ok, so basically, both works but since we don't need the dpath evaluation, we prefer to remove that `*` is that right? 

Knowing this, I'm just wondering about the added complexity of the manifest of having this. If the performance cost is low, I think I would keep the `*` as it seems more maintainable. However, I don't have a strong opinion so I'll let you decide

File: airbyte-integrations/connectors/source-jira/source_jira/utils.py
Comment (darynaishchenko) [reply airbyte-integrations/connectors/source-jira/source_jira/utils.py]: updated with `read_only_records()`, removed custom `_read()`

**Review by maxi297 — COMMENTED** [review]
LGTM! Thanks for the added context that helped me understand the changes. I can't approve as I'm the one that created this PR but I approve

File: airbyte-integrations/connectors/source-jira/source_jira/streams.py
Comment (brianjlai) [inline airbyte-integrations/connectors/source-jira/source_jira/streams.py:178]: hey @darynaishchenko 

Gonna pick this discussion up here since this is presumably where you added the fix to stop the infinite loop. What I'm trying to figure out here is whether there is a bug in the RFR code that is causing the infinite loop or that Jira is configured in a way where we need this fix. 

Right now I feel like this fix you have here would work and it addressed the issue blocking the merge. However, it feels like a workaround to a CDK bug. Ideally we shouldn't have to fix this in the connector

- For RFR streams, we should automatically get the assignment of `self.state` to `FULL_REFRESH_COMPLETE_STATE` within the default implementation of `HttpStream.read_records()` when we finish reading a page [here](https://github.com/airbytehq/airbyte/blob/5056e6782639b05743dc77a9ee731e5f583bae86/airbyte-cdk/python/airbyte_cdk/sources/streams/http/http.py#L408)
- For regular incremental or full refresh streams, those use checkpoint readers (IncrementalCheckpointReader / FullRefreshCheckpointReader) that don't rely on `FULL_REFRESH_COMPLETE_STATE` to stop iteration.

I would be okay if we merged this if we need to get this fix out, but I am going to try to debug on your branch for the `issue_comments` stream to see if I can see what's wrong. I'm okay w/ the merge, but let's add a comment that this is a temporary workaround. I'll keep everyone posted as I know more, but I suspect that we're assigning the wrong checkpoint reader

Comment (darynaishchenko) [reply airbyte-integrations/connectors/source-jira/source_jira/streams.py:178]: Thanks Brian! I'll add comment about this workaround and merge these changes with temporary workaround. I'll run regression tests to be sure that it works fine before merge.

Comment (darynaishchenko) [reply airbyte-integrations/connectors/source-jira/source_jira/streams.py:186]: comment added.
Let me know when you find something.

> but I am going to try to debug on your branch for the issue_comments stream to see if I can see what's wrong.

**Comment by darynaishchenko** [comment]
Regression tests for `issue_comments`, `issue_worklogs`, `issues` streams: [job link](https://github.com/airbytehq/airbyte/actions/runs/10196622465/job/28207746195)
**test_catalog_are_the_same [failed]**: new field `is_resumable`, expected for these changes. Accepted.

**Pull request merged**
